### ISSUE TITLE
Could `BufStream` be flattened?

### ISSUE THREAD

**Issue by aturon** [carllerche/tower-web#111]
As currently defined, `BufStream` produces a stream of `Buf` values, which in turn may contain *multiple* `[u8]` slices. I was curious what the tradeoffs were around, instead, having `BufStream` yield `AsRef<[u8]>` directly, thus "flattening" these two layers. I'm *pretty sure* there are straightforward ways to translate in either direction, but I may be missing something.

**Comment by seanmonstar** [comment]
If I have a buffer and need to provide two slices into it, (say, chunked decoding found a chunked header in the middle), being able to yield the buffer directly as an `impl Buf` where the middle is skipped would be more efficient than cloning the two parts into 2 owned things. Something like `Bytes` allows that to be done cheaply, but even then, it'd be more efficient to only need to `poll` once instead of twice.

Additionally, `Buf` allows people to know how much of the buffer has been "used", via `advance`, in case they want to record as data is written or consumed, or do something like flow control by releasing H2 window capacity.

**Comment by aturon** [comment]
@seanmonstar Yeah, I was imagining using something like `Bytes` to deal with that problem, and was imagining the state machine for polling would have a fast path for the case where you have several slices you're trying to produce. I was also thinking this could be wrapped in a transformation from `Stream<Item = impl Buf>` to `Stream<Item = impl AsRef<[u8]>` or some such. Granted, we could (probably more easily) set things up in the reverse direction; I was mostly looking to see if the core `BufStream` definition could be simplified.

cc @cramertj as well

**Comment by seanmonstar** [comment]
There's not much, but I found https://github.com/hyperium/hyper/issues/1508
